execution/function/function.go

```diff
@@ -0,0 +1,833 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package function
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sort"
+	"sync"
+
+	"github.com/efficientgo/core/errors"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/value"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/thanos-community/promql-engine/execution/model"
+	"github.com/thanos-community/promql-engine/execution/parse"
+)
+
+// functionOperator returns []model.StepVector after processing input with desired function.
+type functionOperator struct {
+	funcExpr  *parser.Call
+	series    []labels.Labels
+	once      sync.Once
+	nextIndex int
+	nextFunc  func(ctx context.Context) ([]model.StepVector, error)
+	nextOps   []model.VectorOperator
+
+	call FunctionCall
+}
+
+func NewfunctionOperator(funcExpr *parser.Call, call FunctionCall, nextOps []model.VectorOperator) model.VectorOperator {
+	f := &functionOperator{
+		nextOps:   nextOps,
+		call:      call,
+		funcExpr:  funcExpr,
+		nextIndex: 0,
+	}
+
+	for i := range funcExpr.Args {
+		if funcExpr.Args[i].Type() == parser.ValueTypeMatrix || funcExpr.Args[i].Type() == parser.ValueTypeVector {
+			f.nextIndex = i
+		}
+	}
+
+	// Decide based on next type.
+	switch funcExpr.Args[f.nextIndex].Type() {
+	case parser.ValueTypeVector, parser.ValueTypeScalar:
+		f.nextFunc = f.instantVectorOrScalarNext
+	case parser.ValueTypeMatrix:
+		f.nextFunc = f.rangeVectorNext
+	default:
+		f.nextFunc = f.defaultNext
```

Comment:
Let's throw the exception here so that we can catch it and fall back.

Reply:
Added error.